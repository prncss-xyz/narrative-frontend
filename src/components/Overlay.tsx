import { useClickOutside } from "../hooks/clickOutside";
import { Box, Flex } from "./basics";

export function Overlay({
  visible,
  setVisible,
  children,
  ...props
}: {
  visible: boolean;
  setVisible: () => void;
  children: React.ReactNode;
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  const ref = useClickOutside(setVisible);
  if (!visible) return null;
  return (
    <>
      <Box
        backgroundColor="black"
        css={{
          position: "fixed",
          left: "0px",
          top: "0px",
          zIndex: 10,
          opacity: "50%",
        }}
        height="100vh"
        width="100vw"
      />
      <Flex
        flexDirection="column"
        css={{
          position: "fixed",
          left: "0px",
          top: "0px",
          zIndex: 10,
        }}
        height="100vh"
        width="100vw"
      >
        <Flex justifyContent="center" height="100%" width="100%">
          <Flex
            justifyContent="center"
            alignItems="center"
            color="text"
            flexDirection="column"
          >
            <Box
              ref={ref}
              backgroundColor="tone1"
              p={2}
              borderStyle="solid"
              {...props}
            >
              {children}
            </Box>
          </Flex>
        </Flex>
      </Flex>
    </>
  );
}
