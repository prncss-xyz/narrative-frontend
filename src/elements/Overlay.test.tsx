import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Overlay } from "./Overlay";

describe("Overlay", () => {
  describe.skip("closed", () => {
    const { container } = render(
      <div data-testid="closed-outside">
        <Overlay
          visible={false}
          onClickOutside={() => {
            /* do nothing */
          }}
        >
          <div>closed-inside</div>
        </Overlay>
      </div>
    );
    it("should not find inside", () => {
      expect(() => {
        screen.getByText("closed-inside");
      }).toThrowErrorMatchingSnapshot('"error"');
    });
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });
  });
  describe("opened", async () => {
    const spy = vi.fn();
    const { container } = render(
      <div data-testid="opened-outside">
        <Overlay visible={true} onClickOutside={spy}>
          <div>opened-inside</div>
        </Overlay>
      </div>
    );
    it("should be visible", () => {
      screen.getByText("opened-inside");
    });
    await userEvent.click(screen.getByTestId("opened-outside"));
    it("should call handler", () => {
      expect(spy).toBeCalledTimes(1);
    });
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });
  });
});
