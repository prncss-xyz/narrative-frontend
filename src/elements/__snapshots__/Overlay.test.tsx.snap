// Vitest Snapshot v1

exports[`Overlay > opened > should match snapshot 1`] = `<div />`;
